index: implement schema management layer for vector search indexes

smoczy123 opened the pull request:
This pull request adds support for creating custom indexes (at a metadata level) as long as a supported custom class is provided (currently only vector search).

The patch contains:

- a change in CREATE INDEX statement that allows for the USING keyword to be present as long as one of the supported classes is used
-  support for describing custom indexes in the DESCRIBE statement
- unit tests 

Co-authored by: @Balwancia

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=23720

denesb commented:
CI Kicked

swasik commented:
Just for reference, the initial review for this was done by @Jadw1 and @piodul in https://github.com/zpp-2024-vector-search/scylladb/pull/3.

nyh commented on test/boost/secondary_index_test.cc:
```diff
@@ -711,11 +711,13 @@ SEASTAR_TEST_CASE(test_secondary_index_create_custom_index) {
-        assert_that_failed(e.execute_cql("create index on cf (a) using 'org.apache.cassandra.index.sasi.SASIIndex'"));
+        // This is the default syntax for specifying a custom index and we 
+        // support "opensearch_vector_index". This should work.
+        e.execute_cql("create custom index on cf (a) using 'opensearch_vector_index'").get();
```
How do we know this command will always work? In the future, we may need to configure an external search engine node, and if we don't, this command will fail, no?

Personally, I think these opensearch tests should be separate test, in their separate file (and perhaps even test/cqlpy, not test/boost).

nyh commented on test/boost/secondary_index_test.cc:
```diff
@@ -711,11 +711,13 @@ SEASTAR_TEST_CASE(test_secondary_index_create_custom_index) {
+        e.execute_cql("create custom index on cf (a) using 'opensearch_vector_index'").get();
+        // It's not a syntax error to try to specify a "USING" without
+        // specifying CUSTOM. This should work.
+        e.execute_cql("create index on cf (b) using 'opensearch_vector_index'").get();
```
Interesting, in the previous code we were sure this was not allowed by Cassandra. And because we wrote this test in test/boost instead of test/cqlpy and couldn't run the test against Cassandra - we didn't know our assumption was wrong.

nyh commented on test/cqlpy/test_describe.py:
```diff
@@ -496,38 +496,60 @@ def is_scylla(cql):
         # Only Scylla supports local indexes
         has_local = is_scylla(cql)
+
+        # Cassandra inserts a space between the table name and parentheses,
```
One of the nice things about the test/cqlpy framework is that there is no major advantage to testing 7 different things in one test function instead of having 7 test functions.

Please leave the original test function unmodified and add a new function testing the  INDEX .. USING (custom index) feature.

Having a separate function makes it easier to understand, gives you a place to add a comment explaining what is being tested and refer to some issue, etc., and if one day we have a regression, we know exactly what failed.

Imagine one day in the future we rename "opensearch_vector_index" and suddenly this test fails. It is nicer to see that the "custom index" test failed but the original index test still passes, and more easily understand what went wrong.

nyh reviewed (changes requested):
Looks mostly good, but I left a few comments and requests. Thanks.

nyh commented on cql3/statements/create_index_statement.cc:
```diff
@@ -348,9 +352,9 @@ std::optional<create_index_statement::base_schema_with_new_index> create_index_s
-        kind = index_metadata_kind::custom;
+    if (_properties->custom_class || _properties->is_custom) {
         index_options = _properties->get_options();
+        kind = index_metadata_kind::custom;
```
I see you did add a test that confirms that dropping the word CUSTOM is fine.
But in this case, I think the is_custom() call in the specific place I commented about here is no longer needed.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16744/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16744/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16744/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/secondary_index_test
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/4742/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/7916/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/15969/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16744/)
#### Build Details:
- Duration: 7 hr 40 min
- Builder: spider4.cloudius-systems.com

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=23720

smoczy123 commented:
Changelog:

- combined two non-test commits
- removed unneeded is_custom check
- simplified index map lookup
- made the supported classes set const static
- moved all of the introduced tests to cqlpy/test_custom_index.py

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=23720

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=23720

avikivity commented on cql3/statements/create_index_statement.cc:
```diff
@@ -67,13 +68,13 @@ create_index_statement::validate(query_processor& qp, const service::client_stat
         throw exceptions::invalid_request_exception("Only CUSTOM indexes can be created without specifying a target column");
     }
-
+    
```
Please don't make gratuitous whitespace changes, especially not ones that add trailing whitespace.

avikivity commented on cql3/statements/create_index_statement.cc:
```diff
@@ -67,13 +68,13 @@ create_index_statement::validate(query_processor& qp, const service::client_stat
 std::vector<::shared_ptr<index_target>> create_index_statement::validate_while_executing(data_dictionary::database db) const {
     auto schema = validation::validate_column_family(db, keyspace(), column_family());
-
+    auto classes = db.find_column_family(schema).get_index_manager().supported_custom_classes(db.features());
```
Seems like the supported classes shouldn't depend on the current table?

nyh replied:
Moreover, as far as I can tell, you don't even use the list of custom classes - you just want to check if a certain custom class is allowed. So it could have been a static function (in the index_manager class) called is_custom_class_supported(name). I'm not sure why db.features() is needed here, but I guess that can be a parameter too...

avikivity commented on cql3/statements/create_index_statement.cc:
```diff
@@ -95,12 +96,25 @@ std::vector<::shared_ptr<index_target>> create_index_statement::validate_while_e
+        for (auto& target : targets) {
+            auto c_def = schema->get_column_definition(to_bytes(target->column_name()));
+            if (c_def->type->get_kind() != abstract_type::kind::vector) {
+                throw exceptions::invalid_request_exception(format("Custom indices currently are only supported for vector columns."));
```
Seems like this check should be performed by the custom index, not here.

swasik replied:
Right, it was also mentioned in [1] so moving the rest of discussion there.

[1] https://github.com/scylladb/scylladb/pull/23720#discussion_r2048445032

avikivity commented on index/secondary_index_manager.cc:
```diff
@@ -341,4 +347,24 @@ bool secondary_index_manager::is_global_index(const schema& s) const {
+    using namespace std::literals;
+    const static std::set<std::string_view> classes = {
+        "usearch_vector_index"sv,
+        "opensearch_vector_index"sv,
```
Maybe we should hide the implementation and just have vector_index here, and have the database configuration (db/config.cc) select the implementation.

avikivity replied:
/cc @tzach

avikivity replied:
What about this?

nyh commented on cql3/statements/create_index_statement.cc:
```diff
@@ -95,12 +96,25 @@ std::vector<::shared_ptr<index_target>> create_index_statement::validate_while_e
+        for (auto& target : targets) {
+            auto c_def = schema->get_column_definition(to_bytes(target->column_name()));
+            if (c_def->type->get_kind() != abstract_type::kind::vector) {
+                throw exceptions::invalid_request_exception(format("Custom indices currently are only supported for vector columns."));
```
This is not exactly true, secondary_index_manager.cc  uses the word "indices" more than indexes:
```
$ grep -c indexes secondary_index_manager.cc
2
$ grep -c indices secondary_index_manager.cc
12
```
Maybe you're saying that we use the form "indexes" in documentation? This I didn't check.

swasik replied:
To be honest my inspiration was system table showing information about indexes that is called `system_schema.indexes`. But out of the curiosity I checked documentation and it is 1120 for indexes vs 41 for indices.

avikivity replied:
Seems like indexes is the preferred form in SQL technical documentation.

nyh commented on test/cqlpy/test_custom_index.py:
```diff
@@ -0,0 +1,75 @@
+###############################################################################
+
+import pytest
+from .util import new_test_table
```
Please import is_scylla() from .util as well, and you won't need to re-implement it below.

nyh commented on test/cqlpy/test_custom_index.py:
```diff
@@ -0,0 +1,75 @@
+# SPDX-License-Identifier: LicenseRef-ScyllaDB-Source-Available-1.0
+
+###############################################################################
+# Tests for custom indexes
```
Nitpick: it looks like this test file will be almost entirely on the the vector search (except just two test functions that I could find), and if we ever have a different type of custom index type, it will get its own tests, so perhaps you should just call this file test_vector_index.py ? 

I am guessing that in the future, you'll also have in this file the ability to skip most of the tests of that file if running on Scylla that is not configured with vector search enabled (remember that vector search needs Scylla to be configured **and** some external search engine needs to run, which not every developer running tests will want to do).

nyh commented on test/cqlpy/test_custom_index.py:
```diff
@@ -0,0 +1,75 @@
+    with new_test_table(cql, test_keyspace, schema) as table:
+        invalid_custom_class = "invalid.custom.class"
+        with pytest.raises(InvalidRequest):
+            cql.execute(f"CREATE CUSTOM INDEX invalid_idx ON {table}(v) USING '{invalid_custom_class}'")
```
Nitpick: you don't need to specify an explict index name ("invalid_idx"). I think it just distraction to the test. I thought for a second that "invalid_idx" was the invalid thing, but it isn't, it's just a random name.

nyh commented on test/cqlpy/test_custom_index.py:
```diff
@@ -0,0 +1,75 @@
+        with pytest.raises(InvalidRequest):
+            cql.execute(f"CREATE CUSTOM INDEX invalid_idx ON {table}(v) USING '{invalid_custom_class}'")
+        with pytest.raises(InvalidRequest):
+            cql.execute(f"CREATE CUSTOM INDEX no_class_idx ON {table}(v)")
```
Please split this second scenario to a separate test, and add a comment explaining what's being tested here - that it not allowed to use the word "CUSTOM" if no "USING" is provided, although as we tested above, the opposite isn't true (you are allowed to use USING without CUSTOM - the CUSTOM is optional and doesn't say anything).

The important benefit of splitting this test is that this test can also be run on Cassandra (see `test/cqlpy/run-cassandra`, and `test/cqlpy/README.md` on how to use if you never did it). This test is supposed to run and pass on Cassandra, while all your other tests that check vector search will not. By the way, this also means you'll need to add `scylla_only` to some of of these tests.

smoczy123 replied:
It seems like Cassandra throws a ConfigurationException here. Should we also throw a ConfigurationException (unlike all other checks in create_index_statement), or differ from Cassandra and stay with InvalidRequest?

nyh replied:
No, I think it's fine to differ in this, I think we have a few of those differences already (and I vaguely remember that Cassandra changed their behavior across versions). If I remember  can support two different exceptions by using a tuple, something like
```
with pytest.raises((InvalidRequest,ConfigurationException)):
```
In one test we even ended up with
```
pytest.raises((InvalidRequest, ConfigurationException, SyntaxException), ...)
```
:-)

nyh commented on test/cqlpy/test_custom_index.py:
```diff
@@ -0,0 +1,75 @@
+def test_create_vector_search_index_on_nonvector_column(cql, test_keyspace):
+    schema = 'p int primary key, v int'
+    with new_test_table(cql, test_keyspace, schema) as table:
+        with pytest.raises(InvalidRequest):
```
Again, please add "match=".  Without it, these  InvalidRequest errors can be anything - syntax error, wrong USING, or whatever...

smoczy123 replied:
The check that this test is related to was removed and this test was changed to xfail. I didn't add match as I don't know what will be the error message from the index check

nyh commented on test/cqlpy/test_custom_index.py:
```diff
@@ -0,0 +1,75 @@
+        
+
+def is_scylla(cql):
+    return any('scylla' in name for name in [row.table_name for row in cql.execute("SELECT * FROM system_schema.tables WHERE keyspace_name = 'system'")])
```
You can import is_scylla from .util instead of having another copy here. @denesb added this function to util.py in 6e53264bc3cf1abcc9621a3e5296c07471c14cd8

smoczy123 commented on cql3/statements/create_index_statement.cc:
```diff
@@ -95,12 +96,25 @@ std::vector<::shared_ptr<index_target>> create_index_statement::validate_while_e
+            throw exceptions::invalid_request_exception(format("Non-supported custom class \'{}\' provided", *(_properties->custom_class)));
+        }
+        
+        for (auto& target : targets) {
```
Should the tests pertaining to this check be removed, commented out or marked as xfail?

swasik replied:
Makes sense to move it to some class implementing vector index. I added #115 having adding validation in scope. 

@smoczy123 I think we can mark tests xfailing with a TODO node as suggested by Nadav saying that this validation should be added to Scylla and reference to #115.

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=23720

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=23720

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=23720

smoczy123 commented:
Changelog:

- Replaced supported_custom_classes with is_custom_class_supported
- Removed check for column type in create_index_statement, added TODO instead
- Added expected reasons for failed tests
- Changed the test related to the deleted check to xfail

smoczy123 commented:
I think it makes sense to use a single vector_index class here. Couldn't we create some kind of an 'internal_vector_index' class in the future if needed?

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=23720

smoczy123 commented:
Changelog:

- Changed the custom classes support to "vector_index" only
- Removed unneeded sv literals
- Removed unneeded escaping in string literals

swasik commented:
@nyh please consider merging

nyh commented:
> @nyh please consider merging

CI didn't run on the latest version... I triggered it myself now, and when it passes we can merge.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17187/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17187/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17187/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/secondary_index_test
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16388/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8307/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5137/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17187/)
#### Build Details:
- Duration: 7 hr 38 min
- Builder: spider2.cloudius-systems.com

swasik commented:
The first run of CI failed on reporting the results back so I re-run it and now we have success.

mykaul commented:
Shouldn't there be some documentation for this feature?

avikivity commented:
> Shouldn't there be some documentation for this feature?

Yes

avikivity commented on cql3/statements/create_index_statement.cc:
```diff
@@ -101,6 +102,14 @@ std::vector<::shared_ptr<index_target>> create_index_statement::validate_while_e
+            throw exceptions::invalid_request_exception(format("Non-supported custom class \'{}\' provided", *(_properties->custom_class)));
+        }
+        
+        //TODO: run custom index class validation once added
```
When will this TODO be done?

smoczy123 replied:
This should be covered by https://github.com/scylladb/vector-store/issues/115

avikivity commented on index/secondary_index_manager.cc:
```diff
@@ -341,4 +346,24 @@ bool secondary_index_manager::is_global_index(const schema& s) const {
+
+    const static std::unordered_set<std::string_view> classes = {
+        "vector_index",
+    };
```
It's better to have this as a map, pointing from the name to the implementation.

Otherwise you'll have to duplicate the name when you look up the implementation.

avikivity commented on test/boost/secondary_index_test.cc:
```diff
@@ -711,11 +711,6 @@ SEASTAR_TEST_CASE(test_secondary_index_create_custom_index) {
-        // specifying CUSTOM. We expect the exception:
-        // "exceptions::invalid_request_exception: Cannot specify index class
-        // for a non-CUSTOM index"
-        assert_that_failed(e.execute_cql("create index on cf (a) using 'org.apache.cassandra.index.sasi.SASIIndex'"));
```
Shouldn't this be part of the previous patch? If you relaxed the rules in the previous patch, then the relaxation of the test should be there too.

But will a SASIIndex creation succeed now? Won't it just fail in a different way?

smoczy123 replied:
It will fail in a different way, however we already test invalid custom class in test_vector_index.py, I don't think there's a reason to duplicate that test.

smoczy123 commented:
> Shouldn't there be some documentation for this feature?

Our documentation already states that 'USING' clause is allowed (see: https://github.com/scylladb/scylladb/blob/5e764d1de28f682939f69ac104084a9f4493402c/docs/cql/secondary-indexes.rst?plain=1#L45C1-L45C83).
We could explicitly mention that we support 'vector_index' custom class, but we neither have the actual implementation yet nor do we support the 'WITH OPTIONS' clause for that class, so if we want to have that in the documentation I think we should wait for the full implementation.

swasik commented:
> > Shouldn't there be some documentation for this feature?
> 
> Our documentation already states that 'USING' clause is allowed (see: https://github.com/scylladb/scylladb/blob/5e764d1de28f682939f69ac104084a9f4493402c/docs/cql/secondary-indexes.rst?plain=1#L45C1-L45C83). We could explicitly mention that we support 'vector_index' custom class, but we don't have the actual implementation yet, so if we want to have that in the documentation I think we should wait for the full implementation. cc @swasik

I created an issue for this (see https://github.com/scylladb/vector-store/issues/136).

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17266/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17266/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17266/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/secondary_index_test
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8370/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5203/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16453/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17266/)
#### Failed Tests (1/44616):
- [test_raft_snapshot_truncation.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17266/testReport/junit/cluster/test_raft_snapshot_truncation/Tests___Unit_Tests___test_raft_snapshot_truncation_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_raft_snapshot_truncation.dev.3)
#### Build Details:
- Duration: 10 hr
- Builder: spider4.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17281/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17281/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17281/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/secondary_index_test
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5218/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8386/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16469/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17281/)
#### Failed Tests (3/44636):
- [test_gcp_provider](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17281/testReport/junit/boost/encryption_at_rest_test/test_gcp_provider) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_gcp_provider)
- [test_gcp_provider_with_impersonated_user](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17281/testReport/junit/boost/encryption_at_rest_test/test_gcp_provider_with_impersonated_user) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_gcp_provider_with_impersonated_user)
- [test_gcp_provider_with_master_key_in_cf](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17281/testReport/junit/boost/encryption_at_rest_test/test_gcp_provider_with_master_key_in_cf) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_gcp_provider_with_master_key_in_cf)
#### Build Details:
- Duration: 7 hr 31 min
- Builder: spider5.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17297/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17297/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17297/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/secondary_index_test
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5232/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8403/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16486/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17297/)
#### Build Details:
- Duration: 6 hr 59 min
- Builder: spider3.cloudius-systems.com

swasik commented:
@nyh please consider merging

nyh commented:
> @nyh please consider merging

Done.

swasik commented:
> @swasik documentation should have been part of the PR, I'm not sure why it's not, let's follow-up with it.

There was already a comment on it somewhere during discussion. We think that it does not make sense to add documentation before we have cloud integration - it will not be usable anyway. We have an issue created in the relevant milestone to remember about it (https://scylladb.atlassian.net/browse/VS-5)
